test/integration/api_paths_test.go

```diff
@@ -0,0 +1,76 @@
+// +build integration
+
+package integration
+
+import (
+	"net/http"
+	"strings"
+	"testing"
+
+	kclient "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/util/sets"
+
+	"github.com/openshift/origin/pkg/cmd/server/origin"
+	testutil "github.com/openshift/origin/test/util"
+	testserver "github.com/openshift/origin/test/util/server"
+)
+
+func TestRootAPIPaths(t *testing.T) {
+	// ExceptionalExpectedCodes are codes that we expect, but are not http.StatusOK.
+	// These codes are expected because the response from a GET on our root should
+	// expose endpoints for discovery, but will not necessarily expose endpoints that
+	// are supported as written - i.e. versioned endpoints or endpoints that need
+	// context will 404 with the correct credentials and that is OK.
+	ExceptionalExpectedCodes := map[string]int{
+		"/api/v1beta3":   http.StatusNotFound,
+		"/logs/":         http.StatusNotFound,
+		"/osapi/v1beta3": http.StatusNotFound,
+	}
+
+	masterConfig, adminConfigFile, err := testserver.StartTestMaster()
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	clientConfig, err := testutil.GetClusterAdminClientConfig(adminConfigFile)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	transport, err := kclient.TransportFor(clientConfig)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	// We need to make sure that any APILevels specified in the config are present in the RootPaths, and that
+	// any not specified are not
+	expectedLevels := sets.NewString(masterConfig.APILevels...)
+	actualLevels := sets.String{}
+	for _, route := range origin.RootPaths.Paths {
+		if strings.HasPrefix(route, "/api/") {
+			actualLevels.Insert(route[5:])
+		}
+	}
+	if !expectedLevels.Equal(actualLevels) {
+		t.Errorf("actual API levels served don't match expected levels:\n\texpected:\n\t%s\n\tgot:\n\t%s", expectedLevels.List(), actualLevels.List())
```

Comment:
"second" integration test that makes sure we only serve the levels we are asked to here @deads2k